libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -362,6 +365,77 @@ struct NoVariablesInInterfaceChecker: public PostTypeChecker::Checker
 	/// Flag indicating whether we are currently inside a StructDefinition.
 	int m_insideStruct = 0;
 };
+
+struct CircularContractDependencies: public PostTypeChecker::Checker
+{
+	CircularContractDependencies(ErrorReporter& _errorReporter):
+		Checker(_errorReporter)
+	{}
+
+	bool visit(ContractDefinition const& _contract) override
+	{
+		auto joined = boost::join(
+			_contract.annotation().creationCallGraph->get()->createdContracts,
+			_contract.annotation().deployedCallGraph->get()->createdContracts
+		);
+
+		for (auto const& [contract, referencee]: joined)
+			if (contract != &_contract)
+				_contract.annotation().contractDependencies.emplace(contract, referencee);
+
+		return true;
+	}
+
+	void endVisit(ContractDefinition const& _contract) override
+	{
+		auto const& [contract, referencee] = dependenciesAreCyclic(_contract);
+
+		if (contract)
+			m_foundCycles.emplace(contract, referencee);
+	}
+
+	void finalize() override
+	{
+		for (auto const& [cycle, referencee]: m_foundCycles)
+		{
+			SecondarySourceLocation ssl{};
+			ssl.append(string{"Referenced contract"}, cycle->location());
+
+			m_errorReporter.typeError(
+				7813_error,
+				referencee->location(),
+				ssl,
+				"Circular reference found."
+			);
+		}
+	}
+
+private:
+	map<ContractDefinition const*, ASTNode const*, ASTNode::CompareByID> m_foundCycles;
+
+
+	pair<ContractDefinition const*, ASTNode const*> dependenciesAreCyclic(
+		ContractDefinition const& _currentContract
+	) const
+	{
+		auto visitor = [&](ContractDefinition const& _contract, util::CycleDetector<ContractDefinition>& _cycleDetector, size_t _depth)
+		{
+			if (_depth >= 256)
+				m_errorReporter.fatalTypeError(7864_error, _contract.location(), "Contract dependencies exhausting cyclic dependency validator");
+
+			for (auto& [contract, referencee]: _contract.annotation().contractDependencies)
```

Comment:
I'd differentiate these names a bit. When I notice `_contract` and `contract` used next to each other I always have to go back and check what each of them meant.

```suggestion
			for (auto& [dependencyContract, referencee]: _contract.annotation().contractDependencies)
```